File: src/main/java/leetcode/p017/Solution.java
package leetcode.p017;

import java.util.*;

class Solution {
    public List<String> letterCombinations(String digits) {
        LinkedList<String> combinations = new LinkedList<>();
        recurse(combinations,"",0,digits);
        return combinations;
    }

    void recurse(LinkedList<String> sofar, String cur, int pos, String digits){
        if (pos == digits.length()){//we parsed the full string
            StringBuilder sb = new StringBuilder();
            sofar.add(cur);
        } else {
            Character focus = digits.charAt(pos);
            switch (focus) {
                case '2':
                    recurse(sofar,cur + 'a',pos + 1,digits);
                    recurse(sofar,cur + 'b',pos + 1,digits);
                    recurse(sofar,cur + 'c',pos + 1,digits);
                    break;
                case '3':
                    recurse(sofar,cur + 'd',pos + 1,digits);
                    recurse(sofar,cur + 'e',pos + 1,digits);
                    recurse(sofar,cur + 'f',pos + 1,digits);
                    break;
                case '4':
                    recurse(sofar,cur + 'g',pos + 1,digits);
                    recurse(sofar,cur + 'h',pos + 1,digits);
                    recurse(sofar,cur + 'i',pos + 1,digits);
                    break;
                case '5':
                    recurse(sofar,cur + 'j',pos + 1,digits);
                    recurse(sofar,cur + 'k',pos + 1,digits);
                    recurse(sofar,cur + 'l',pos + 1,digits);
                    break;
                case '6':
                    recurse(sofar,cur + 'm',pos + 1,digits);
                    recurse(sofar,cur + 'n',pos + 1,digits);
                    recurse(sofar,cur + 'o',pos + 1,digits);
                    break;
                case '7':
                    recurse(sofar,cur + 'p',pos + 1,digits);
                    recurse(sofar,cur + 'q',pos + 1,digits);
                    recurse(sofar,cur + 'r',pos + 1,digits);
                    recurse(sofar,cur + 's',pos + 1,digits);
                    break;
                case '8':
                    recurse(sofar,cur + 't',pos + 1,digits);
                    recurse(sofar,cur + 'u',pos + 1,digits);
                    recurse(sofar,cur + 'v',pos + 1,digits);
                    break;
                case '9':
                    recurse(sofar,cur + 'v',pos + 1,digits);
                    recurse(sofar,cur + 'w',pos + 1,digits);
                    recurse(sofar,cur + 'x',pos + 1,digits);
                    recurse(sofar,cur + 'y',pos + 1,digits);
                    recurse(sofar,cur + 'z',pos + 1,digits);
                    break;
            }
        }

    }
}

File: scala/src/leetcode/p017/Solution.scala
package leetcode.p017

object Solution {
    def letterCombinations(digits: String): List[String] = {
        
        if (digits == "") 
            return List.empty[String]

        val mapping: Map[Char, List[Char]] = Map(
            ('2' -> List('a','b', 'c')),
            ('3' -> List('d','e', 'f')),
            ('4' -> List('g','h', 'i')),
            ('5' -> List('j','k', 'l')),
            ('7' -> List('p','q', 'r', 's')),
            ('8' -> List('t','u', 'v')),
            ('9' -> List('w','x', 'y', 'z')),
            ('6' -> List('m','n', 'o')),
        )

        var remaindigits = digits.toList
        rec(digits.toList, mapping, "" , List.empty[String])    
    }

    def rec( remaindigits: List[Char], mapping: Map[Char, List[Char]], sofar: String, res: List[String]): List[String] = {
       if (remaindigits.isEmpty){
        return sofar +: res
       } 
       val curdigit = remaindigits.head
       var letters = mapping(curdigit)
       var newres = res 
       while (!letters.isEmpty){
            val curletter = letters.head 
            newres = rec(remaindigits.tail, mapping, s"${sofar}${curletter}", newres ) 
            letters = letters.tail
       }
       newres
    }
}
